shibboleth_login/auth_backend.py

```diff
@@ -35,62 +23,138 @@ def authenticate(self, request, shibd_meta=None):
+                user = qs.first()
+
+        # create missing users
+        if not user and app_settings.ALLOW_CREATE_NEW_USERS:
+            logger.info('Creating a new Shibboleth authenticated user: %s <%s>',
+                username, email)
+            user = UserModel(**{
+                username_field: username,
+                email_field: email or '',
+            })
+            if not email:
+                user.save()
+                # TODO: use real domain with support for this and pseudonymized users
+                user.email = '{:d}@localhost'.format(user.id)
             user.set_unusable_password()
             user_save_flag = True
-        else:
-            email = self._parse(shibd_meta, MAIL_KEY, 254)
-            if email and email != user.email:
-                user.email = email
-                user_save_flag = True
 
-        first_name = self._parse(shibd_meta, FIRST_NAME_KEY, 30)
+        if not user:
+            return None
+
+        # update email
+        if email and email != user.email:
+            user.email = email
+            user_save_flag = True
+
+        # update first_name
+        first_name = ' '.join(parser.get_values(env_settings.FIRST_NAME, ''))[:first_name_len]
         if first_name and first_name != user.first_name:
             user.first_name = first_name
             user_save_flag = True
 
-        last_name = self._parse(shibd_meta, LAST_NAME_KEY, 30)
+        # update last_name
+        last_name = ' '.join(parser.get_values(env_settings.LAST_NAME, ''))[:last_name_len]
         if last_name and last_name != user.last_name:
             user.last_name = last_name
             user_save_flag = True
 
+        # if not first_name or last_name, fallback to cn and displayName
+        if not user.first_name or not user.last_name:
+            # best effort to find best possible name..
+            full_name = max((
+                ' '.join(parser.get_values(env_settings.FULL_NAME, '')),
+                ' '.join(parser.get_values(env_settings.COMMON_NAME, '')),
+            ), key=len)
+            first_, __, last_ = full_name.partition(' ')
+            if not user.first_name:
+                user.first_name = first_[:first_name_len]
+            if not user.last_name:
+                user.last_name = last_[:last_name_len]
+            user_save_flag = True
+
         if user_save_flag:
+            # TODO: write better error reporting, when there is a validator to raise something
+            user.clean()
             user.save()
 
+        # TODO: support multiple domains
         profile = user.userprofile
-        student_id = self._parse(shibd_meta, STUDENT_ID_KEY)
-        if student_id:
-            student_id = student_id.split(':')[-1][:25]
-            if student_id and student_id != profile.student_id:
-                profile.student_id = student_id
-                profile.save()
+        sid_filters = env_settings.STUDENT_FILTERS.copy()
```

Comment:
Just to check, does the parser for student ids work with non-integer values? Some student ids mix digits and letters. The long Shibboleth URN seems to include the word `int` even when the value really isn't an integer.

Reply:
I think the `int` refers to international in this case. I thought about integers first too.